Enables the huggingface checkpoint conversion to MaxText orbax.

wang2yn84 opened the pull request:
# Description

This PR converts the huggingface llama checkpoint to MaxText orbax format. The purpose of the PR is to convert the Deepseek distilled checkpoint to MaxText. The original llama_or_mistral_ckpt.py only works on the Pytorch checkpoint, but not the Huggingface checkpoint. This PR fixes the work flow and documented the whole process.

Right now the accuracy still has some issue and need further debug.

# Tests

Run through the workflow multiple times and the generated checkpoint works. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

singh-mitali reviewed (commented):
Many lint-only changes in this PR. Can those be skipped. Perhaps run "bash code_style.sh" under maxtext dir

wang2yn84 commented:
> Many lint-only changes in this PR. Can those be skipped. Perhaps run "bash code_style.sh" under maxtext dir

True! Updated and removed those changes.

anfals commented:
@richjames0 had this PR: https://github.com/AI-Hypercomputer/maxtext/pull/1028 to support safetensors, but looks like this is doing more?

wang2yn84 commented:
> @richjames0 had this PR: #1028 to support safetensors, but looks like this is doing more?

Yes I'm aware of this PR, but it doesn't work when I try that conversion function. More logics are required to handle the discrepancies between Huggingface model structure and MaxText. That's why I have this PR.

anfals commented:
> > @richjames0 had this PR: #1028 to support safetensors, but looks like this is doing more?
> 
> Yes I'm aware of this PR, but it doesn't work when I try that conversion function. More logics are required to handle the discrepancies between Huggingface model structure and MaxText. That's why I have this PR.

Gotcha! Yeah I was pointed to this older PR for converting a HF chpt, but we are seeing major issues with loss as I run with it. Your comment more or less confirms the ckpt conversion was the problem. But it'll be good once this PR lands and merges

singh-mitali commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -192,12 +215,296 @@ def __getitem__(self, key):
+  return x
+
+
+def convert_huggingface_to_jax_weights(base_model_path, model_size, huggingface_ckpt, model_params, mem_info):
```
This function and the following old function look very similar except the loading function. Could they be combined?

singh-mitali reviewed (approved):
Left some comments - but those could be addresses in a follow up CL.

RissyRan commented:
Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.

wang2yn84 commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -168,6 +168,29 @@ def _hf_mapping(layer_idx: int = -1, expert_idx: int = -1) -> dict:
   }
 
 
+def _hf_to_maxtext_mapping(layer_idx: int = -1, expert_idx: int = -1) -> dict:
```
I'm actually considering deleting that mapping cuz it's not working. Will refactor in the follow up PRs.

wang2yn84 commented:
> Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.

Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.

RissyRan commented on MaxText/tests/__init__.py:
```diff
@@ -0,0 +1,15 @@
+"""
+Copyright 2023 Google LLC
```
Nit: 2025

RissyRan commented on MaxText/tests/forward_pass_logit_checker.py:
```diff
@@ -77,7 +77,7 @@ def get_data(golden_data, golden_data_index, config):
 
 
-def main(config, test_args):
+def main(config, test_args):  # pylint: disable=W0621
```
Will `bash code_style.sh` work for those?

wang2yn84 replied:
Yea I used bash code_style.sh and it report test_args is redefined within main function. That's weird and functionally everything is working fine. So I have to disable it for now.

RissyRan replied:
Oh.... test_args shouldn't be redefined and overwritten. Probably we shouldn't disable it, but find the cause?

wang2yn84 replied:
It's not redefined anywhere. That's the myth.

RissyRan commented:
> > Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.
> 
> Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.

Yes, it works. @richjames0 and I were working on those ckpt from huggingface. To be specific we tested:

1) download safetensors from https://huggingface.co/mistralai/Mixtral-8x22B-v0.1
2) run script with this [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028/files)

What's the issue you met? Key not finding?

vipannalla commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -192,12 +215,242 @@ def __getitem__(self, key):
+    wk = np.reshape(wk, [base_num_query_heads * head_dim, base_num_kv_heads, head_dim])
+    wv = np.reshape(wv, [base_num_query_heads * head_dim, base_num_kv_heads, head_dim])
+
+    if model_size[:8] == "llama3.1":
```
Is this logic applicable for `3.1` version or all version after `3.1` as well?

wang2yn84 replied:
We only have 3.1 for now. Yes it should work for say, 3.3

RissyRan replied:
Is this change needed later? update it to 3.3?

vipannalla commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -192,12 +215,242 @@ def __getitem__(self, key):
+    self_attention["value"]["kernel"][layer_idx, ...] = wv  # pylint: disable=E1137
+    self_attention["out"]["kernel"][layer_idx, ...] = w_post  # pylint: disable=E1137
+
+  self_attention["query"]["kernel"] = np.transpose(self_attention["query"]["kernel"], axes=(1, 0, 2, 3))
```
Nit: can you add a comment about the hardcoded axes `(1, 0, 2, 3)` and what they refer to in maxtext/Jax?

wang2yn84 replied:
Done.

vipannalla commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -192,12 +215,242 @@ def __getitem__(self, key):
+    layer_weight["post_self_attention_layer_norm"]["scale"][layer_idx, ...] = post_self_attention_layernorm  # pylint: disable=E1137
+
+  layer_weight["pre_self_attention_layer_norm"]["scale"] = np.transpose(
+      layer_weight["pre_self_attention_layer_norm"]["scale"], axes=(1, 0)
```
Please add comment about what `(1, 0)` axes refer to here.

vipannalla commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -192,12 +215,242 @@ def __getitem__(self, key):
 
+  if num_experts is None:
+    # swap the layer index
+    layer_weight["mlp"]["wi_0"]["kernel"] = np.transpose(layer_weight["mlp"]["wi_0"]["kernel"], axes=(1, 0, 2))
```
please add doc about hardcoded axes for posterity...

vipannalla commented on MaxText/llama_or_mistral_ckpt.py:
```diff
@@ -219,12 +472,13 @@ def convert_to_jax_weights(base_model_path, model_size):
   max_logging.log(f"Loading the base model from {base_model_path}")
   # Skip any hidden files for checkpoints
-  ckpt_paths = sorted(pathlib.Path(base_model_path).glob("[!.]*.pth"))
+  if huggingface_ckpt:
+    return _convert_huggingface_to_jax_weights(base_model_path, model_size, model_params, mem_info)
```
This is probably for future PRs, but can we also consolidate the rest of the logic into `_convert_pytorch_to_jax_weights()` to make it cleaner?

wang2yn84 replied:
Good point! Updated!

vipannalla reviewed (commented):
Looks great, thanks Lance for adding this script. I had few minor comments.

wang2yn84 commented:
> > Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.
> 
> Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.



> > > Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.
> > 
> > 
> > Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.
> 
> Yes, it works. @richjames0 and I were working on those ckpt from huggingface. To be specific we tested:
> 
> 1. download safetensors from https://huggingface.co/mistralai/Mixtral-8x22B-v0.1
> 2. run script with this [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028/files)
> 
> What's the issue you met? Key not finding?

As far as I can see, llama_or_mistral_ckpt.py doesn't have the safetensor loader and it can only load pth file. How do you load safetensor checkpoint?

RissyRan commented:
Are 

> > > Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.
> > 
> > 
> > Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.
> 
> > > > Thanks for the change! The old [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028) works fine on my side a while ago, I'd like to take a look at this PR.
> > > 
> > > 
> > > Does it converts the Huggingface checkpoint? The name mapping has the wrong direction, it's from MaxText to Huggingface. I would be surprised if it worked before.
> > 
> > 
> > Yes, it works. @richjames0 and I were working on those ckpt from huggingface. To be specific we tested:
> > 
> > 1. download safetensors from https://huggingface.co/mistralai/Mixtral-8x22B-v0.1
> > 2. run script with this [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028/files)
> > 
> > What's the issue you met? Key not finding?
> 
> As far as I can see, llama_or_mistral_ckpt.py doesn't have the safetensor loader and it can only load pth file. How do you load safetensor checkpoint?

You should be able to find this one in https://github.com/AI-Hypercomputer/maxtext/pull/1028/files

```
def load_safetensors_checkpoint(ckpt_paths):
  chkpt_vars_raw = {}
  for i, ckpt_path in enumerate(ckpt_paths):
    max_logging.log(f"Loading checkpoint path {i+1} of {len(ckpt_paths)} ...")
    with safe_open(ckpt_path, framework="pt") as f:
      for k in f.keys():
        assert k not in chkpt_vars_raw
        chkpt_vars_raw[k] = f.get_tensor(k)
  chkpt_vars = [_HFNamespaceMapper(chkpt_vars_raw)]
  return chkpt_vars
```

Sorry that we haven't merged this [PR](https://github.com/AI-Hypercomputer/maxtext/pull/1028/files) in time due to a minor comment. Please don't merge until we are aligned. Due the urgency, I am ok to save this in a branch or copy of a separate file.

RissyRan commented on MaxText/tests/hf_checkpoint_conversion_checker.py:
```diff
@@ -0,0 +1,156 @@
+
+def test_huggingface_to_maxtext_back_to_huggingface_flow():
+  base_num_query_heads = 16
+  head_dim = 32
```
Why those 2 config are defined/hardcode in the test?

wang2yn84 replied:
I moved this test to a separate file. Basically this is a unit test for the permutation function. So everything else is hardcoded.

wang2yn84 commented on MaxText/configs/base.yml:
```diff
@@ -518,6 +519,7 @@ inference_metadata_file: "" # path to a json file
 inference_server: "MaxtextInterleavedServer"  # inference server to start
 inference_benchmark_test: False
 enable_model_warmup: False
+hf_model_path: ""  # inference checkpoint correctness verification
```
When I run the script, if I don't add it to base.yml, it complains it's configured in the command line but not in the config.

RissyRan replied:
Interesting! How come? as `max_kl_div`, `atol`, etc are not in the base.yml as well.

wang2yn84 replied:
Discussed offline and agreed to removed it from here. It's excluded in the test. Should use "--" to pass in the config.
